libsolidity/experimental/ast/TypeSystem.cpp

```diff
@@ -0,0 +1,348 @@
+{
+	std::vector<UnificationFailure> failures;
+	auto unificationFailure = [&]() {
+		failures.emplace_back(UnificationFailure{TypeMismatch{_a, _b}});
+	};
+	_a = resolve(_a);
+	_b = resolve(_b);
+	std::visit(util::GenericVisitor{
+		[&](TypeVariable _left, TypeVariable _right) {
+			if (_left.index() == _right.index())
+			{
+				if (_left.sort() != _right.sort())
+					unificationFailure();
+			}
+			else
+			{
+				if (_left.sort() <= _right.sort())
+					failures += instantiate(_left, _right);
+				else if (_right.sort() <= _left.sort())
+					failures += instantiate(_right, _left);
+				else
+				{
+					Type newVar = m_typeSystem.freshVariable(_left.sort() + _right.sort());
+					failures += instantiate(_left, newVar);
+					failures += instantiate(_right, newVar);
+				}
+			}
+		},
+		[&](TypeVariable _var, auto) {
+			failures += instantiate(_var, _b);
+		},
+		[&](auto, TypeVariable _var) {
+			failures += instantiate(_var, _a);
+		},
+		[&](TypeConstant _left, TypeConstant _right) {
+			if (_left.constructor != _right.constructor)
+				return unificationFailure();
+			if (_left.arguments.size() != _right.arguments.size())
+				return unificationFailure();
+			for (auto&& [left, right]: ranges::zip_view(_left.arguments, _right.arguments))
+				failures += unify(left, right);
+		},
+		[&](auto, auto) {
+			unificationFailure();
+		}
+	}, _a, _b);
+	return failures;
+}
+
+bool TypeEnvironment::typeEquals(Type _lhs, Type _rhs) const
+{
+	return std::visit(util::GenericVisitor{
+		[&](TypeVariable _left, TypeVariable _right) {
+			if (_left.index() == _right.index())
+			{
+				solAssert(_left.sort() == _right.sort());
+				return true;
+			}
+			return false;
+		},
+		[&](TypeConstant _left, TypeConstant _right) {
+			if (_left.constructor != _right.constructor)
+				return false;
+			if (_left.arguments.size() != _right.arguments.size())
+				return false;
+			for (auto&& [left, right]: ranges::zip_view(_left.arguments, _right.arguments))
+				if (!typeEquals(left, right))
+					return false;
+			return true;
+		},
+		[&](auto, auto) {
+			return false;
+		}
+	}, resolve(_lhs), resolve(_rhs));
+}
+
+TypeEnvironment TypeEnvironment::clone() const
+{
+	TypeEnvironment result{m_typeSystem};
+	result.m_typeVariables = m_typeVariables;
+	return result;
+}
+
+TypeSystem::TypeSystem()
+{
+	auto declarePrimitiveClass = [&](std::string _name) {
+		return std::visit(util::GenericVisitor{
+			[](std::string _error) -> TypeClass {
+				solAssert(false, _error);
+			},
+			[](TypeClass _class) -> TypeClass { return _class; }
+		}, declareTypeClass(freshVariable({}), _name, nullptr));
+	};
+
+	m_primitiveTypeClasses.emplace(PrimitiveClass::Type, declarePrimitiveClass("type"));
+
+	for (auto [type, name, arity]: std::initializer_list<std::tuple<PrimitiveType, char const*, size_t>>{
+		{PrimitiveType::TypeFunction, "tfun", 2},
+		{PrimitiveType::Function, "fun", 2},
+		{PrimitiveType::Function, "itself", 1},
```

Comment:
```suggestion
		{PrimitiveType::Itself, "itself", 1},
```

Reply:
Strange, actually, that that doesn't cause issues.

Reply:
Just ran into it myself and came here to point it out. And I see it has already been reported :)

I'm pushing a fixup to correct this.

Reply:
We need to be more careful in rebasing this then - this was already merged before :-).

Reply:
I'm pretty sure I just accepted the suggestion via github - but we will need to occasionally rebase this - actually also now, since the basic infrastructure PR is merged.